toc/rfc/rfc-draft-v2-app-manifests.md

```diff
@@ -0,0 +1,613 @@
+
+Example manifest:
+```yaml
+---
+apps:
+  - name: app1
+    lifecycle:
+      type: buildpack
+      buildpacks:
+        - name: my_buildpack
+          url: git.example.com/my_buildpack
+```
+
+#### Shared Resources
+
+Some resources can be shared across multiple spaces (currently, routes and
+service instances). Sharing resources is not currently possible via app
+manifests. Since these resources are no longer fully-contained within a space,
+they require configuration split across multiple space-level manifests.
+
+One option could be to add a boolean `shared` node to sharable resources, so
+they are not automatically created in the destination space, but can still be
+bound/mapped/etc to.
+
+Source manifest:
+```yaml
+---
+routes:
+  - url: route.example.com
+    destinations:
+      - app:
+          name: app1
+service_instances:
+  - name: my-service-instance
+    bindings:
+      apps:
+        - name: app1
+    share_to:
+      - org: my-org
+        space: my-space
+```
+
+Destination manifest:
+```yaml
+---
+routes:
+  - url: route.example.com
+    shared: true
+    destinations:
+      - app:
+          name: app1
+service_instances:
+  - name: my-service-instance
+    shared: true
+    bindings:
+      apps:
+        - name: app1
+```
+
+#### Network Policies
+
+Similar to shared resources, container networking policies can span multiple
+spaces. Currently, it is not possible to configure network policies via app
+manifests.
+
+Notably, network polices are managed by Policy Server, not by Cloud Controller,
+so an additional integration will likely need to be built to support network
+policies in manifests.
+
+Manifest support for network policies is an [oft-requested
+feature](https://github.com/cloudfoundry/cloud_controller_ng/issues/1334) and
+will be required for implementing GitOps workflows for apps that use container
+networking.
+
+Destination manifest:
+```yaml
+---
+routes:
+  - url: destination-app.apps.internal
+    destinations:
+      - app:
+          name: destination-app
+network_policies:
+  - destination:
+      app:
+        name: destination-app
+        port_range: 1234-9999
+    source:
+      app:
+        name: source-app
+        org: source-app-org
+        space: source-app-space
+    protocol: tcp
+```
+
+Network policies will be specified in the destination app's space. No
+configuration will be needed in the source app's space.
+
+#### Per-Route Options
+
+The recent [per-route features
+RFC](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0027-generic-per-route-features.md)
+included additions to the manifest.
+
+Example manifest:
+```yaml
+---
+routes:
+  - url: route.example.com/path
+    options:
+      loadbalancing-algorithm: round-robin
+      connection-limit: 15
+      session-cookie: FOOBAR
+      trim-path: true
+```
+
+Note that the options are intentionally arbitrary and opaque to Cloud
+Controller, hence why hyphenated keys are tolerated.
```

Comment:
I took these examples directly from [RFC 0027](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0027-generic-per-route-features.md#cloud-controller). I'm happy with whatever keys the networking group decides on.

There are also other places in the manifest where we accept arbitrary keys, where we won't enforce conformity. For example:
- Arbitrary parameters for service instances and service bindings
- Metadata labels and annotations
- Environment variables

Reply:
@a18e / @hoffmaen correct me if I'm wrong but IIRC we decided to go with whatever the common pattern is in the manifest. They are not truly arbitrary after all since the CC has to validate them.

Maybe we should update the other RFC as well to reflect that.

Reply:
Ah, I think I misread the original RFC. I thought that CC wasn't responsible for validating these options. Looking that the associated CC PRs, I see that isn't true. I'll update this section of the example.

Reply:
Addressed by 91c214ce92677dcfff96b5f144f3b984e30d5d57